internal/app/init/pkg/system/runner/containerd/containerd.go

```diff
@@ -6,147 +6,207 @@ package containerd
+type containerdRunner struct {
+	data *userdata.UserData
+	args *runner.Args
+	opts *runner.Options
 
-// WithMemoryLimit sets the linux resource memory limit field.
-func WithMemoryLimit(limit int64) oci.SpecOpts {
-	return func(_ context.Context, _ oci.Client, _ *containers.Container, s *specs.Spec) error {
-		s.Linux.Resources.Memory = &specs.LinuxMemory{
-			Limit: &limit,
-			// DisableOOMKiller: &disable,
-		}
-		return nil
-	}
+	stop    chan struct{}
+	stopped chan struct{}
 }
 
-// WithRootfsPropagation sets the root filesystem propagation.
-func WithRootfsPropagation(rp string) oci.SpecOpts {
-	return func(_ context.Context, _ oci.Client, _ *containers.Container, s *specs.Spec) error {
-		s.Linux.RootfsPropagation = rp
-		return nil
+// errStopped is used internally to signal that task was stopped
+var errStopped = errors.New("stopped")
+
+// NewRunner creates runner.Runner that runs a container in containerd
+func NewRunner(data *userdata.UserData, args *runner.Args, setters ...runner.Option) runner.Runner {
+	r := &containerdRunner{
+		data:    data,
+		args:    args,
+		opts:    runner.DefaultOptions(),
+		stop:    make(chan struct{}),
+		stopped: make(chan struct{}),
 	}
+
+	for _, setter := range setters {
+		setter(r.opts)
+	}
+
+	return r
 }
 
 // Run implements the Runner interface.
 // nolint: gocyclo
-func (c *Containerd) Run(data *userdata.UserData, args runner.Args, setters ...runner.Option) error {
-	// Wait for the containerd socket.
+func (c *containerdRunner) Run() error {
+	defer close(c.stopped)
 
+	// Wait for the containerd socket.
 	_, err := conditions.WaitForFileToExist(defaults.DefaultAddress)()
 	if err != nil {
 		return err
 	}
 
-	// Create the default runner options.
-
-	opts := runner.DefaultOptions()
-	for _, setter := range setters {
-		setter(opts)
-	}
-
 	// Create the containerd client.
 
-	ctx := namespaces.WithNamespace(context.Background(), opts.Namespace)
+	ctx := namespaces.WithNamespace(context.Background(), c.opts.Namespace)
 	client, err := containerd.New(defaults.DefaultAddress)
 	if err != nil {
 		return err
 	}
 	// nolint: errcheck
 	defer client.Close()
 
-	image, err := client.GetImage(ctx, opts.ContainerImage)
+	image, err := client.GetImage(ctx, c.opts.ContainerImage)
 	if err != nil {
 		return err
 	}
 
 	// Create the container.
 
-	specOpts := newOCISpecOpts(image, args, opts)
-	containerOpts := newContainerOpts(image, args, opts, specOpts)
+	specOpts := c.newOCISpecOpts(image)
+	containerOpts := c.newContainerOpts(image, specOpts)
 	container, err := client.NewContainer(
 		ctx,
-		args.ID,
+		c.args.ID,
 		containerOpts...,
 	)
 	if err != nil {
-		return fmt.Errorf("failed to create container %q: %v", args.ID, err)
+		return errors.Wrapf(err, "failed to create container %q", c.args.ID)
+	}
+	defer container.Delete(ctx, containerd.WithSnapshotCleanup) // nolint: errcheck
+
+	// Manage task lifecycle
+	switch c.opts.Type {
+	case runner.Once:
```

Comment:
there's one inconsistency I haven't fixed and left it the way it was: `Once` for process means "restart until succeeds", while `Once` for containerd means "run exactly once"